Sources/NIOPosix/NIOThreadPool.swift

```diff
@@ -58,17 +57,46 @@ public final class NIOThreadPool {
 
     /// The work that should be done by the `NIOThreadPool`.
     public typealias WorkItem = @Sendable (WorkItemState) -> Void
+
+    /// `WorkItem`-s encapsulated in a reference type in order to avoid CoW when mutating the queue.
+    /// Since `WorkItems` are embedded in `State` enum, it's hard to reliably avoid CoW otherwise.
+    ///
+    /// Also the closures are wrapped in a struct, to avoid the cost of allocation as discussed on https://bugs.swift.org/browse/SR-15872
+    final class WorkItems {
```

Comment:
Instead of boxing the buffer here with a class we should be able to avoid the CoW by introducing a new state called `modifying` into the `State` enum and before mutating the queue we transition to that state and then back to the running state. We have used that approach in most other places recently and it avoids this class entirely. 

The other upside of this pattern is that once we language gets `mutating/inout` switches in the future we can easily transition over to that and remove the `modifying` case again.

Reply:
good catch! shoot, missed the `class`, yes, let's do this properly

Reply:
The `.modifying` solution is 60% slower than the `class`. @FranzBusch @weissi do you still want me to do that?

Reply:
Where does the slow down come from and is sprinkling some `@inlinable` + `@usableFromInline` making it go faster?

Reply:
Ah, this confused me also.

Turns out I had a different testing environment. When I was developing the original solution, and performance testing it. I was sitting on top of #2645, which also has an effect, as ELFs are used to notify completions.

In the "this is 60% slower" result, I was sitting on top of now rebased branch from this PR.

My bad. I'll update the results in the PR description to give the numbers for this branch alone (as they should have been). And then we'll see the 60% perf bump on once #2645 goes in.

Reply:
@dnadoba I checked `Deque` and it didn't have any noticeable performance difference. Is it supposed to? And/or are we attempting to move away from `CircularBuffer`? Pushing `Deque`-s version, so that we have all the different paints on the PR :).

Reply:
We are moving away from `CircularBuffer`. We are using `Deque` in all new code e.g. in our new custom AsyncSequences. 
Both are heavily optimised but `Deque` is the way forward. This isn't urgent and we will keep `CircularBuffer` around but it isn't likely to be updated. All new performance optimisations, if any are possible, should be done on top of `Deque` as it is independent of NIO and therefore more generally useful.

Reply:
@dnadoba got it, thanks. Will keep this in mind for the future. This code could be trivially switched to `Deque` so it's already in.